### ISSUE TITLE
The NF does not perform unit checking on arrays or elements thereof

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#5685]

**Comment by casella** [comment]
@perost, please check the following MWE
```modelica
partial model M
  parameter Integer N = 2;
  Modelica.SIunits.MassFlowRate w[N];
  Modelica.SIunits.Volume V[N];
  Modelica.SIunits.Density rho[N];
equation 
  rho[1]*der(V[1]) = w[1]; // Correct equation
  der(V[1]) = w[1];        // Dimensionally inconsistent
  der(V) = w;              // Dimensionally inconsistent
  for i in 1:N loop
    der(V[i]) = w[i];      // Dimensionally inconsistent
  end for;  
end M;
``` 
I'm afraid something went wrong since you closed [the original ticket](https://trac.openmodelica.org/OpenModelica/ticket/5685) three years ago, so array variables are no longer considered by unit checking. Maybe because we switched to the NF?

**Comment by perost** [comment]
@casella: The issue should be fixed in #9446.

The actual issue was that a while back I changed how we handle type attributes such that for array variables we actually get the correct array expression, so now we get e.g. `Real x[3](unit = fill("m3", 3))` rather than `Real x[3](unit = "m3")` that we could get previously. However, the unit checking was only designed to handle scalar literal string expressions, so it would fail on this and just ignore the unit.

**Comment by casella** [comment]
Great, thanks @perost for the quick fix!

@matteodepascali, this will be available with the next nightly build.
